beacon_chain/state/state_transition.py

```diff
@@ -484,10 +488,77 @@ def calculate_crosslink_rewards(crystallized_state: CrystallizedState,
     validators = crystallized_state.validators
     rewards_and_penalties = [0 for _ in validators]  # type: List[int]
 
-    #
-    # STUB
-    # Still need clarity in spec to properly fill these calculations
-    #
+    total_deposits = crystallized_state.total_deposits
+    total_deposits_in_ETH = total_deposits // WEI_PER_ETH
+    reward_quotient = config['base_reward_quotient'] * int(sqrt(total_deposits_in_ETH))
+    quadratic_penalty_quotient = int(sqrt(config['sqrt_e_drop_time'] / config['slot_duration']))
+
+    last_state_recalc = crystallized_state.last_state_recalc
+
+    # collect crosslink participation data for each shard_id that was attempted to
+    # be crosslinked two cycles ago
+    committee_crosslinks = {}  # type: Dict[str, Any]
+    for slot in range(max(last_state_recalc - config['cycle_length'], 0), last_state_recalc):
+        shards_and_committees = get_shards_and_committees_for_slot(
+            crystallized_state,
+            slot,
+            config=config
+        )
+        for shard_and_committee in shards_and_committees:
+            shard_id = shard_and_committee.shard_id
+            if shard_id not in committee_crosslinks:
+                committee_crosslinks[shard_id] = {
+                    'participating_validator_indices': [],
+                    'non_participating_validator_indices': [],
+                    'total_participated_v_deposits': 0,
+                    'total_v_deposits': 0
+                }
+            attestations = [
+                attestation for attestation in active_state.pending_attestations
+                if attestation.slot == slot and attestation.shard_id == shard_id
+            ]
+            if attestations:
+                bitfields = [
+                    attestation.attester_bitfield for attestation in attestations
+                ]
+                bitfield = or_bitfields(bitfields)
+            else:
+                bitfield = get_empty_bitfield(len(shard_and_committee.committee))
+
+            committee_crosslink = committee_crosslinks[shard_id]
+            for i, index in enumerate(shard_and_committee.committee):
```

Comment:
IMO `i` and `index` are confusing. What do you think of renaming `index` to `validator_index`?

Reply:
going to go with `i` and `validator_index`.

Reply:
actually, I went with `committee_index, validator_index`